File: presentation/src/main/java/com/marvelcomics/brito/presentation/ResourceModel.java
package com.marvelcomics.brito.presentation;

public class ResourceModel<T> {

    private State state;
    private T data;
    private String message;
    private Throwable throwable;

    public ResourceModel(State state, T data, String message) {
        this.state = state;
        this.data = data;
        this.message = message;
    }

    public ResourceModel(State state, T data, String message, Throwable throwable) {
        this.state = state;
        this.data = data;
        this.message = message;
        this.throwable = throwable;
    }

    public State getState() {
        return state;
    }

    public T getData() {
        return data;
    }

    public String getMessage() {
        return message;
    }

    public Throwable getThrowable() {
        return throwable;
    }

    public enum State {
        LOADING, SUCCESS, ERROR
    }
}

File: data/src/main/java/com/marvelcomics/brito/data/handler/ResourceModel.kt
package com.marvelcomics.brito.data.handler

data class ResourceModel<out Type>(val status: State, val data: Type?, val message: String?) {
    companion object {
        fun <Type> success(data: Type?): ResourceModel<Type> {
            return ResourceModel(State.SUCCESS, data, null)
        }

        fun <Type> error(statusCode: Int, message: String, data: Type?): ResourceModel<Type> {
            return ResourceModel(State.ERROR, data, message)
        }

        fun <Type> loading(data: Type? = null): ResourceModel<Type> {
            return ResourceModel(State.LOADING, data, null)
        }
    }

    enum class State {
        LOADING, SUCCESS, ERROR, DEFAULT
    }
}